### PR TITLE
spirv-reduce: Support reducing a specific function

### PR DESCRIPTION
**Pull request by afd**
Motivated by integrating spirv-reduce into spirv-fuzz, so that an
added function can be made smaller during shrinking, this adds support
in spirv-reduce for asking reduction to be restricted to the
instructions of a single specified function.

### CODE REVIEW

File: tools/reduce/reduce.cpp [hunk 1]
```diff
@@ -169,6 +178,15 @@ ReduceStatus ParseFlags(int argc, const char** argv,
...
         reducer_options->set_step_limit(step_limit);
+      } else if (0 == strncmp(cur_arg, "--target-function=",
+                              sizeof("--target-function=") - 1)) {
+        const auto split_flag = spvtools::utils::SplitFlagArgs(cur_arg);
+        char* end = nullptr;
+        errno = 0;
+        const auto target_function =
+            static_cast<uint32_t>(strtol(split_flag.second.c_str(), &end, 10));
+        assert(end != split_flag.second.c_str() && errno == 0);
+        reducer_options->set_target_function(target_function);
```
Comment (paulthomson) [inline tools/reduce/reduce.cpp:189]: Could alternatively do something like this (for input `temp`). This is from gf-layers.

```
  std::istringstream ss{temp};
  std::uint64_t temp_uint{};
  ss >> temp_uint;

  if (ss.fail()) {
    LOG("Failed to parse %s", temp);
    return false;
  }

  *value = temp_uint;
  return true;
```

**Review by paulthomson — APPROVED** [review]
LGTM!

File: source/reduce/reduction_opportunity_finder.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      result.push_back(&function);
+    }
+  }
+  assert((!target_function || !result.empty()) &&
+         "Requested target function must exist.");
+  return result;
+}
+
+}  // namespace reduce
+}  // namespace spvtools
```
[the remark this answers is not in the document]
Comment (afd) [reply source/reduce/reduction_opportunity_finder.cpp:34]: Fixed.

File: source/reduce/remove_unused_struct_member_reduction_opportunity_finder.cpp [hunk 3]
```diff
@@ -24,7 +24,14 @@ namespace reduce {
 
 std::vector<std::unique_ptr<ReductionOpportunity>>
 RemoveUnusedStructMemberReductionOpportunityFinder::GetAvailableOpportunities(
-    opt::IRContext* context) const {
+    opt::IRContext* context, uint32_t target_function) const {
+  if (target_function) {
+    // Removing an unused struct member is a global change, a struct types are
+    // global.  We thus do not consider such opportunities if we are targeting
+    // a specific function.
```
[the remark this answers is not in the document]
Comment (afd) [reply source/reduce/remove_unused_struct_member_reduction_opportunity_finder.cpp:31]: Fixed, thanks.

File: tools/reduce/reduce.cpp (hunk 1 above)
Comment (afd) [reply tools/reduce/reduce.cpp:189]: This looks good, but I think we should do it in a separate PR and for all the options.

**Pull request merged**